File: crates/client/src/facades/sync.rs
use std::collections::BTreeSet;

use ipc::errors::IpcError;

use crate::session::Session;

pub struct Sync<'a> {
    pub(crate) session: &'a Session,
}

impl Sync<'_> {
    pub async fn get(&self) -> Result<proto::sync::SyncConfig, IpcError> {
        self.session
            .call::<proto::sync::SyncGet>(&proto::Empty {})
            .await
    }

    pub async fn sources(
        &self,
        unit: proto::sync::SyncUnit,
    ) -> Result<Vec<proto::sync::SyncSource>, IpcError> {
        let params = proto::sync::SyncSourcesParams { unit };
        Ok(self
            .session
            .call::<proto::sync::SyncSources>(&params)
            .await?
            .sources)
    }

    /// `source` may be empty only when at most one instance holds the file.
    pub async fn enable(
        &self,
        unit: proto::sync::SyncUnit,
        source: &str,
    ) -> Result<proto::sync::SyncConfig, IpcError> {
        let params = proto::sync::SyncEnableParams {
            unit,
            source: source.to_string(),
        };
        self.session.call::<proto::sync::SyncEnable>(&params).await
    }

    pub async fn disable(
        &self,
        unit: proto::sync::SyncUnit,
    ) -> Result<proto::sync::SyncConfig, IpcError> {
        let params = proto::sync::SyncDisableParams { unit };
        self.session.call::<proto::sync::SyncDisable>(&params).await
    }

    pub async fn set_unsynced(
        &self,
        unsynced: BTreeSet<String>,
    ) -> Result<proto::sync::SyncConfig, IpcError> {
        let params = proto::sync::SyncKeysParams { unsynced };
        self.session.call::<proto::sync::SyncKeys>(&params).await
    }

    pub async fn options(&self) -> Result<Vec<proto::sync::SyncOption>, IpcError> {
        Ok(self
            .session
            .call::<proto::sync::SyncOptionsGet>(&proto::Empty {})
            .await?
            .options)
    }

    pub async fn set_option(
        &self,
        key: &str,
        value: &str,
    ) -> Result<Vec<proto::sync::SyncOption>, IpcError> {
        let params = proto::sync::SyncOptionSetParams {
            key: key.to_string(),
            value: value.to_string(),
        };
        Ok(self
            .session
            .call::<proto::sync::SyncOptionSet>(&params)
            .await?
            .options)
    }

    pub async fn status(&self) -> Result<Vec<proto::sync::InstanceSyncStatus>, IpcError> {
        Ok(self
            .session
            .call::<proto::sync::SyncStatus>(&proto::Empty {})
            .await?
            .instances)
    }

    /// `shared` unset returns the instance to following the catalogue.
    pub async fn set_instance_unit(
        &self,
        instance: &str,
        unit: proto::sync::SyncUnit,
        shared: Option<bool>,
    ) -> Result<proto::sync::InstanceSyncStatus, IpcError> {
        let params = proto::sync::InstanceSyncUnitParams {
            instance: instance.to_string(),
            unit,
            shared,
        };
        self.session
            .call::<proto::sync::InstanceSyncUnit>(&params)
            .await
    }

    pub async fn set_instance_unsynced(
        &self,
        instance: &str,
        unsynced: BTreeSet<String>,
    ) -> Result<proto::sync::InstanceSyncStatus, IpcError> {
        let params = proto::sync::InstanceSyncKeysParams {
            instance: instance.to_string(),
            unsynced,
        };
        self.session
            .call::<proto::sync::InstanceSyncKeys>(&params)
            .await
    }
}

File: frontend/src/api/sync.ts
/**
 * The `sync.*` channels — the catalogue of shared settings, what can seed it,
 * the shared game options, and each instance's standing.
 */
import { call } from './core/ipc';
import type {
  InstanceSyncStatus,
  SyncConfig,
  SyncOption,
  SyncSource,
  SyncUnit,
} from './types/sync';

export function get(): Promise<SyncConfig> {
  return call('sync.get');
}

/** Which instances could start a unit's shared copy. */
export async function sources(unit: SyncUnit): Promise<SyncSource[]> {
  const result = await call<{ sources: SyncSource[] }>('sync.sources', {
    unit,
  });
  return result.sources;
}

/** `source` may be empty only when at most one instance holds the file. */
export function enable(unit: SyncUnit, source = ''): Promise<SyncConfig> {
  return call('sync.enable', { unit, source });
}

export function disable(unit: SyncUnit): Promise<SyncConfig> {
  return call('sync.disable', { unit });
}

/** The `options.txt` keys no instance shares. */
export function setUnsynced(unsynced: string[]): Promise<SyncConfig> {
  return call('sync.options.keys', { unsynced });
}

export async function options(): Promise<SyncOption[]> {
  const result = await call<{ options: SyncOption[] }>('sync.options.get');
  return result.options;
}

export async function setOption(
  key: string,
  value: string,
): Promise<SyncOption[]> {
  const result = await call<{ options: SyncOption[] }>('sync.options.set', {
    key,
    value,
  });
  return result.options;
}

export async function status(): Promise<InstanceSyncStatus[]> {
  const result = await call<{ instances: InstanceSyncStatus[] }>('sync.status');
  return result.instances;
}

/** `shared` null returns the instance to following the catalogue. */
export function setInstanceUnit(
  instance: string,
  unit: SyncUnit,
  shared: boolean | null,
): Promise<InstanceSyncStatus> {
  return call('instance.sync.unit', { instance, unit, shared });
}

export function setInstanceUnsynced(
  instance: string,
  unsynced: string[],
): Promise<InstanceSyncStatus> {
  return call('instance.sync.keys', { instance, unsynced });
}
